node/src/services/blockchain/transaction.rs

```diff
@@ -0,0 +1,150 @@
+use std::{
+    str::FromStr,
+    time::{Duration, Instant},
+};
+
+use anyhow::anyhow;
+use log::*;
+use sp_core::H256;
+use storage_hub_infra::actor::ActorHandle;
+use tokio::sync::mpsc::Receiver;
+
+use crate::services::blockchain::{commands::BlockchainServiceInterface, types::ExtrinsicResult};
+
+use super::{types::ExtrinsicHash, BlockchainService};
+
+const LOG_TARGET: &str = "blockchain-transaction";
+
+#[derive(Debug)]
+pub struct SubmittedTransaction {
+    watcher: Receiver<String>,
+    hash: ExtrinsicHash,
+    timeout: Option<Duration>,
+}
+
+const NO_TIMEOUT_WARN: Duration = Duration::from_secs(60);
+
+impl SubmittedTransaction {
+    pub fn new(watcher: Receiver<String>, hash: H256) -> Self {
+        Self {
+            watcher,
+            hash,
+            timeout: None,
+        }
+    }
+
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    pub async fn watch_for_success(
+        &mut self,
+        blockchain: &ActorHandle<BlockchainService>,
+    ) -> anyhow::Result<()> {
+        // Wait for the transaction to be included in a block.
+        let block_hash;
+
+        let start_time = Instant::now();
+        loop {
+            // Get the elapsed time since submit.
+            let elapsed = start_time.elapsed();
+            // Calculate the remaining time to wait.
+            let remaining = match self.timeout {
+                Some(timeout) => {
+                    // Check if the timeout has been reached.
+                    if elapsed > timeout {
+                        return Err(anyhow!(
+                            "Timeout waiting for transaction to be included in a block"
```

Comment:
Maybe add the tx hash to this error message to help with identifying the context.

Reply:
Agreed, it'd be useful to add it for all error messages.